README
NAME
    GitStore - Git as versioned data store in Perl

VERSION
    version 0.16

SYNOPSIS
        use GitStore;

        my $gs = GitStore->new('/path/to/repo');
        $gs->set( 'users/obj.txt', $obj );
        $gs->set( ['config', 'wiki.txt'], { hash_ref => 1 } );
        $gs->commit();
        $gs->set( 'yyy/xxx.log', 'Log me' );
        $gs->discard();
    
        # later or in another pl
        my $val = $gs->get( 'user/obj.txt' ); # $val is the same as $obj
        my $val = $gs->get( 'config/wiki.txt' ); # $val is { hashref => 1 } );
        my $val = $gs->get( ['yyy', 'xxx.log' ] ); # $val is undef since discard

DESCRIPTION
    It is inspired by the Python and Ruby binding. check SEE ALSO

CLASS FUNCTIONS
  create( $directory )
    Creates the directory, initialize it as a git repository, and returns
    its associated "GitStore" object.

        my $gs = GitStore->create( '/tmp/mystore' );

METHODS
  new
        GitStore->new('/path/to/repo');
        GitStore->new( repo => '/path/to/repo', branch => 'mybranch' );
        GitStore->new( repo => '/path/to/repo', author => 'Someone Unknown <unknown\@what.com>' );

    repo
        your git directory or work directory (*GitStore* will assume it's a
        work directory if it doesn't end with ".git").

    branch
        your branch name, default is 'master'

    author
        It is used in the commit info

    serializer
        Can be used to define a serializing function that will be used if
        the value to save is a reference. When invoked, the function will be
        passed a reference to the store object, the path under which the
        value will be saved, and the value itself. For example, one could do
        different serialization via:

            my $store = GitStore->new(
                repo => '/path/to/repo',
                serializer => sub {
                    my( $store, $path, $value ) = @_;

                    if ( $path =~ m#^json# ) {
                        return encode_json($value);
                    }
                    else {
                        # defaults to YAML
                        return YAML::Dump($value);
                    }
                },
            );

        The default serializer uses "nfreeze" in Storable.

    deserializer
        Called when a value is picked from the store to be (potentially)
        deserialized. Just like the serializer function, it is passed three
        arguments: the store object, the path of the value to deserialize
        and the value itself. To revisit the example for "serializer", the
        full serializer/deserializer dance would be:

            my $store = GitStore->new(
                repo => '/path/to/repo',
                serializer => sub {
                    my( $store, $path, $value ) = @_;

                    return $path =~ m#^json# 
                                        ? encode_json($value)
                                        : YAML::Dump($value)
                                        ;
                },
                deserializer => sub {
                    my( $store, $path, $value ) = @_;
            
                    return $path =~ #^json#
                                        ?decode_json($value)
                                        : YAML::Load($value)
                                        ;
                },
            );

        The default deserializer will try to deserialize the value retrieved
        from the store via "thaw" in Storable and, if this fails, return the
        value verbatim.

    autocommit
        If set to "true", any call to "set()" or "delete()" will
        automatically call an implicit "commit()". Defaults to "false".

  set($path, $val)
        $gs->set( 'yyy/xxx.log', 'Log me' );
        $gs->set( ['config', 'wiki.txt'], { hash_ref => 1 } );
        $gs->set( 'users/obj.txt', $obj );

    Store $val as a $path file in Git

    $path can be String or ArrayRef. Any leading slashes ('/') in the path
    will be stripped, as to make it a valid Git path. The same grooming is
    done for the "get()" and "delete()" methods.

    $val can be String or Ref[HashRef|ArrayRef|Ref[Ref]] or blessed Object

  get($path)
        $gs->get( 'user/obj.txt' );
        $gs->get( ['yyy', 'xxx.log' ] );

    Get $val from the $path file

    $path can be String or ArrayRef

  get_revision( $path )
    Like "get()", but returns the GitStore::Revision object corresponding to
    the latest Git revision on the monitored branch for which $path changed.

  delete($path)
  remove($path)
    remove $path from Git store

  commit
        $gs->commit();
        $gs->commit('Your Comments Here');

    commit the set changes into Git

  discard
        $gs->discard();

    discard the set changes

  exist($path)
    Returns *true* if an object exists at the given path.

  history($path)
    Returns a list of GitStore::Revision objects representing the changes
    brought to the *$path*. The changes are returned in ascending commit
    order.

  list($regex)
        @entries = $gs->list( qr/\.txt$/ );

    Returns a list of all entries in the repository, possibly filtered by
    the optional *$regex*.

FAQ
  why the files are not there?
    run

        git checkout

  any example?
        # if you just need a local repo, that's all you need.
        mkdir sandbox
        cd sandbox
        git init
        # use GitStore->new('/path/to/this/sandbox')
        # set something
        git checkout
    
        # follows are for remote git url
        git remote add origin git@github.com:fayland/sandbox2.git
        git push origin master
        # do more GitStore->new('/path/to/this/sandbox') later
        git checkout
        git pull origin master
        git push

KNOWN BUGS
    If all files are deleted from the repository, a 'dummy' file will be
    created to keep Git happy.

SEE ALSO
    Article
        <http://www.newartisans.com/2008/05/using-git-as-a-versioned-data-st
        ore-in-python.html>

    Python binding
        <http://github.com/jwiegley/git-issues/tree/master>

    Ruby binding
        <http://github.com/georgi/git_store/tree/master>

Git URL
    <http://github.com/fayland/perl-git-store/tree/master>

AUTHORS
    *   Fayland Lam <fayland@gmail.com>

    *   Yanick Champoux <yanick@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2015 by Fayland Lam <fayland@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for GitStore

{{$NEXT}}
 [API CHANGES]

 [BUG FIXES]

 [DOCUMENTATION]

 [ENHANCEMENTS]

 [NEW FEATURES]

 [STATISTICS]

0.17 2015-02-08
 [BUG FIXES]
 - Tweak test as cloning was reported as a failure.

 [STATISTICS]
 - code churn: 2 files changed, 12 insertions(+), 5 deletions(-)

0.16 2015-02-07
 [ENHANCEMENTS]
 - add exist() method.
 - add create() function.

 [STATISTICS]
 - code churn: 11 files changed, 146 insertions(+), 64 deletions(-)

0.15 2013-05-16
 [ENHANCEMENTS]
 - add MooseX::Storage::IO::GitStore

 [STATISTICS]
 - code churn: 4 files changed, 160 insertions(+), 3 deletions(-)

0.14 2013-05-16
 [ENHANCEMENTS]
 - serialization is now configurable.
 - new 'autocommit' attribute.

 [STATISTICS]
 - code churn: 7 files changed, 214 insertions(+), 26 deletions(-)

0.13 2012-11-11
 [NEW FEATURES]
 - New get_revision() method.

 [STATISTICS]
 - code churn: 4 files changed, 67 insertions(+), 7 deletions(-)

0.12 2012-11-04
 [NEW FEATURES]
 - New list() method to list entries.

0.11 2012-06-24
 [BUG FIXES]
 - files with directories where saved as filename '/foo/bar.txt' instead of
   file 'bar.txt' in a directory 'foo'. Caused problems if dealing with a
   generic git repository.

 [STATISTICS]
 - code churn: 7 files changed, 236 insertions(+), 84 deletions(-)

0.10 2012-06-17
 [BUG FIXES]
 - Prevents GitStore from dying if all files are deleted from the
   repository.
 - Leading '/' in files passed to 'set()', 'get()' and 'delete()' are
   removed, as they are not valid Git paths.
   https://github.com/fayland/perl-git-store/issues/2, thanks to Chris
   Forbes for reporting.

0.09 2012-06-07
 [ENHANCEMENTS]
 - Ported all uses of Git::Repository to Git::PurePerl

0.08 2012-06-05
 - Add history() method.

0.07 2010-07-23
 - fix the commits such that they don't clobber their branch anymore. 

0.06 2010-03-03
 - Change the way author is handled to work with new version of
   Git::PurePerl::NewObject::Commit.

0.05 2009-03-16
 - fix commit date by time()
 - fix commit author/committer by configurable author

0.04 2009-03-16
 - FAQ doc

0.03 2009-03-14
 - big bug fix because I used Git::PurePerl wrong, thanks for acme

0.02 2009-03-14
 - fix broken CPAN dist

0.01 2009-03-14
 - First version, released on an unsuspecting world.



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Module::Build" => "0.3601",
    "Test::More" => "0.88"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Git as versioned data store in Perl",
  "dist_author" => [
    "Fayland Lam <fayland\@gmail.com>",
    "Yanick Champoux <yanick\@cpan.org>"
  ],
  "dist_name" => "GitStore",
  "dist_version" => "0.15",
  "license" => "perl",
  "module_name" => "GitStore",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "DateTime" => 0,
    "Git::PurePerl" => 0,
    "List::Util" => 0,
    "Moose" => 0,
    "Moose::Role" => 0,
    "Moose::Util::TypeConstraints" => 0,
    "Path::Class" => 0,
    "Storable" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/GitStore/Revision.pm
package GitStore::Revision;
BEGIN {
  $GitStore::Revision::AUTHORITY = 'cpan:YANICK';
}
#ABSTRACT: the state of a given path for a specific commit
$GitStore::Revision::VERSION = '0.16';

use strict;
use warnings;

use Moose;

use GitStore;
use DateTime;
use List::Util qw/ first /;
use Path::Class;


has sha1 => (
    is => 'ro',
    required => 1,
);




has commit_object => (
    is => 'ro',
    lazy => 1,
    default => sub {
        my $self = shift;

        return $self->git->get_object($self->sha1);
    },
    handles => {
        timestamp => 'authored_time',
        message   => 'comment',
    },
);


has path => (
    is => 'ro',
    required => 1,
);

has gitstore => (
    is => 'ro',
    isa => 'GitStore',
    required => 1,
    handles => {
        git_repo => 'git_repo',
        git => 'git',
    },
);

has file_object => ( 
    is => 'ro',
    lazy => 1,
    default => sub { 
        my $self = shift;

        $self->gitstore->_find_file( 
            $self->commit_object->tree, file($self->path) 
        );
    },
);



sub content {
    my $self = shift;

    $self->gitstore->deserializer->($self->gitstore,$self->path,$self->file_object->object->content);
}


__PACKAGE__->meta->make_immutable;
1;

__END__

=pod

=encoding UTF-8

=head1 NAME

GitStore::Revision - the state of a given path for a specific commit

=head1 VERSION

version 0.16

=head1 SYNOPSIS

   use GitStore;

   my $gs = GitStore->new('/path/to/repo');

   my @history = $gs->history( 'path/to/object' );

   for my $rev ( @history ) {
        say "modified at: ", $rev->timestamp;
        say "commit message was: ", $rev->message;
        say "===\n", $rev->content;
   }

=head1 DESCRIPTION

Represents an object in a  L<GitStore> at a specific commit.

=head1 METHODS

=head2 sha1

Returns the SHA-1 of the commit.

=head2 commit_object

Returns the L<Git::PurePerl::Object::Commit> object containing the file revision.

=head2 timestamp

Returns the commit time of the revision as a L<DateTime> object.

=head2 message

Returns the commit message of the revision.  Note that the message might have
additional trailing carriage returns.

=head2 path

Returns the path of the L<GitStore> object.

=head2 content

Returns the content of the object.  If the object is a frozen ref, the
structure will be returned, like for `GitStore`'s `get()`.

=head1 AUTHORS

=over 4

=item *

Fayland Lam <fayland@gmail.com>

=item *

Yanick Champoux <yanick@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Fayland Lam <fayland@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/MooseX/Storage/IO/GitStore.pm
package MooseX::Storage::IO::GitStore;
BEGIN {
  $MooseX::Storage::IO::GitStore::AUTHORITY = 'cpan:YANICK';
}
#ABSTRACT:  Save Moose objects in your GitStore
$MooseX::Storage::IO::GitStore::VERSION = '0.16';
use Moose::Role;

requires 'thaw';
requires 'freeze';

has git_repo => (
    traits => [ 'DoNotSerialize' ],
    is => 'ro',
    isa => 'Str',
    required => 0,
);

has gitstore => (
    traits => [ 'DoNotSerialize' ],
    is => 'ro',
    isa => 'GitStore|Undef',
    lazy => 1,
    default => sub { $_[0]->git_repo && _basic_store($_[0]->git_repo); },
);

sub _basic_store {
    GitStore->new( repo => $_[0], autocommit => 1, 
        serializer => sub { $_[2] },
        deserializer => sub { $_[2] }, 
    );
}

sub load {
    my ( $class, $filename, %args ) = @_;

    if( ref $class ) {
        $args{git_repo} ||= $class->git_repo;
        $args{gitstore} ||= $class->gitstore;
    }

    $args{inject}{git_repo} ||= $args{git_repo};
    $args{gitstore} ||= _basic_store($args{git_repo});
    $class->thaw($args{gitstore}->get($filename),%args);
}

sub store {
    my ( $self, $filename, %args ) = @_;
    ($self->gitstore || _basic_store($args{git_repo}))->set( $filename => $self->freeze(%args) );
}

no Moose::Role;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

MooseX::Storage::IO::GitStore - Save Moose objects in your GitStore

=head1 VERSION

version 0.16

=head1 SYNOPSIS

  package Point;
  use Moose;
  use MooseX::Storage;

  with Storage( format => 'YAML', io => 'GitStore');

  has 'x' => (is => 'rw', isa => 'Int');
  has 'y' => (is => 'rw', isa => 'Int');

  1;

  my $p = Point->new(x => 10, y => 10);

  ## methods to load/store a class
  ## on the file system

  $p->store('my_point', git_repo => './store');

  my $p2 = Point->load('my_point', git_repo => './store');

=head1 ATTRIBUTES

The role adds two attributes to the class, C<git_repo> and C<gitstore>.

=head2 git_repo

The optional path to the git repository where the object will be stored.

=head2 gitstore

The L<GitStore> object used by the storage system. If not specified, it will
be created using I<git_repo>, and will have I<autocommit> set to 1.

=head1 METHODS

=over 4

=item B<load ($filename, git_repo => $path)>

Load the object associated with I<$filename>. I<$path> can be omited if used 
from an object instead of as a class method.

=item B<store ($filename, git_repo => $path)>

Save the object as I<$filename>. I<$path> can be omited if used 
from an object instead of as a class method.

=back

=head1 SEE ALSO

L<MooseX::Storage>

=head1 AUTHORS

=over 4

=item *

Fayland Lam <fayland@gmail.com>

=item *

Yanick Champoux <yanick@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Fayland Lam <fayland@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.031.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Git as versioned data store in Perl",
  "AUTHOR" => "Fayland Lam <fayland\@gmail.com>, Yanick Champoux <yanick\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "GitStore",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "GitStore",
  "PREREQ_PM" => {
    "Carp" => 0,
    "DateTime" => 0,
    "Git::PurePerl" => 0,
    "List::Util" => 0,
    "Moose" => 0,
    "Moose::Role" => 0,
    "Moose::Util::TypeConstraints" => 0,
    "Path::Class" => 0,
    "Path::Tiny" => 0,
    "Storable" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Exporter" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.88",
    "lib" => 0,
    "parent" => 0
  },
  "VERSION" => "0.16",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "DateTime" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "FindBin" => 0,
  "Git::PurePerl" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "List::Util" => 0,
  "Moose" => 0,
  "Moose::Role" => 0,
  "Moose::Util::TypeConstraints" => 0,
  "Path::Class" => 0,
  "Path::Tiny" => 0,
  "Storable" => 0,
  "Test::More" => "0.88",
  "lib" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



